### PR TITLE
feat(Auth): Implementation of fetchAuthSession

### PR DESCRIPTION
**Pull request by royjit**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthenticationProviderAdapter.swift [hunk 1]
```diff
@@ -15,4 +15,8 @@ class AuthenticationProviderAdapter: AuthenticationProviderBehavior {
     init(awsMobileClient: AWSMobileClientBehavior) {
         self.awsMobileClient = awsMobileClient
     }
+
+    func isSignedIn() -> Bool {
+        true //return awsMobileClient.isSignedIn()
```
Comment (palpatim) [inline AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthenticationProviderAdapter.swift:20]: What does this comment mean? Is it a TODO, old behavior that needs to be removed, or what?

File: AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthorizationProviderAdapter+SignedInSession.swift [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        // Call getTokens first to verify that the signedIn user is in UserPool or not.
+        awsMobileClient.getTokens { [weak self] tokens, error in
+            guard let self = self else {
+                return
+            }
+            guard error == nil else {
+
+                if let urlError = error as NSError?, urlError.domain == NSURLErrorDomain {
+                    let authSession = AuthCognitoSignedInSessionHelper.offlineSignedInSession()
+                    completionHandler(.success(authSession))
```
Comment (palpatim) [inline AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthorizationProviderAdapter+SignedInSession.swift:41]: See if you can break the handlers for these various error conditions into separate methods, to help with the readability of this method. Right now, it's very hard to follow why you make the decisions you do, and some well-named methods would go a long way toward making your intent clearer. Ideally you'd be cascading these decisions in easy-to-consume chunks, although please do balance the practical considerations of dealing with underlying AWSMobileClient patterns against the need for clarity in this internal module--don't go to extraordinary lengths to clean up code if it means weird contortions to pass values through various callbacks.

File: AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthorizationProviderAdapter+SignedInSession.swift [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    /// - User is signedIn through Cognito User Pool but aws credentials couldnot be refreshed because of N/W or
+    /// service issue
+    /// - User is signedIn through Cognito Identity Pool
+    ///- User is signedIn through Cognito Identity Pool but aws credentials couldnot be refreshed because of N/W or
+    ///service issue
+    /// - Parameter completionHandler: Completion handler to return the result.
+    func fetchSignedInSession( _ completionHandler: @escaping (Result<AuthSession, AmplifyAuthError>) -> Void) {
+
+        // User can be signed in to user pools or federated to identityPool.
+        // Call getTokens first to verify that the signedIn user is in UserPool or not.
```
Comment (palpatim) [inline AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthorizationProviderAdapter+SignedInSession.swift:32]: Note this as an explicit precondition in the doc comments, that this must be called only after verifying that the user is not in `guest` or `signedOut` states, and that the operation is happening in a thread-safe way

File: AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthorizationProviderAdapter+SignedInSession.swift [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+extension AuthorizationProviderAdapter {
+
+    /// Creates session for a signedIn user.
+    ///
+    /// Session values available as a signedIn user depend on the way the user signedIn and the configurations enabled
+    /// for the current signedIn user. The following scenarios can happen:
+    /// - User is signedIn through Cognito User Pool and Identity Pool is not configured
+    /// - User is signedIn through Cognito User Pool but refresh token is expired
+    /// - User is signedIn through Cognito User Pool but access token couldnot be refreshed because of N/W or
```
Comment (palpatim) [inline AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthorizationProviderAdapter+SignedInSession.swift:21]: nit: "N/W" -> "network"

File: AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthorizationProviderAdapter+SignedInSession.swift [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+                let authError = AuthErrorHelper.toAmplifyAuthError(error!)
+                let authSession = AuthCognitoSignedInSessionHelper.signedInSession(withUnhandledError: authError)
+                completionHandler(.success(authSession))
+                return
+            }
+
+            guard let tokens = tokens else {
+                // This should not happen, return an unknown error.
+                let error = AmplifyAuthError.unknown("""
+                    Could not fetch the cognito tokens
```
Comment (palpatim) [inline AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthorizationProviderAdapter+SignedInSession.swift:75]: Put some more context here: we reported the state/session/user as signedIn, but we can't retrieve Cognito Identity credentials, or user pool tokens--describe the situation more fully. I'm not saying we'll ever hit this error, but when we do, we can save ourselves a lot of effort if we clearly identify the expectations going into it.

File: AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthorizationProviderAdapter+SignedInSession.swift [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+                let authError = AuthErrorHelper.toAmplifyAuthError(error!)
+                let authSession = AuthCognitoSignedInSessionHelper.signedInSession(withUnhandledError: authError)
+                completionHandler(.success(authSession))
+                return
+            }
+
+            guard let credentials = awsCredentials else {
+                // This should not happen, return an unknown error.
+                let error = AmplifyAuthError.unknown("""
+                    Could not fetch the AWS Credentials
```
Comment (palpatim) [inline AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthorizationProviderAdapter+SignedInSession.swift:145]: as above, be explicit about the expectations -- why we believe this shouldn't happen

File: AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthorizationProviderAdapter+SignedOutSession.swift [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+                let authSession = AuthCognitoSignedOutSessionHelper.signedOutSession(withError: error!)
+                completionHandler(.success(authSession))
+                return
+            }
+
+            guard let credentials = awsCredentials else {
+                // This should not happen, throw an unknown error.
+                // Since we couldnot fetch aws credentials, we stop here to avoid sending partial information.
+                let error = AmplifyAuthError.unknown("""
+                    Could not fetch the credentials
```
Comment (palpatim) [inline AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthorizationProviderAdapter+SignedOutSession.swift:27]: As elsewhere, let's spell out our expectations: why do we believe this shouldn't happen?

File: AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AWSCredentials+Extension.swift [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+import AWSCore
+
+extension AWSCredentials {
+
+    func toAmplifyAWSCredentials() -> AuthAWSCognitoCredentials {
+        // Credentials are fetched through Cognito Identity Pool and thus these are temporary credentials
```
Comment (palpatim) [inline AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AWSCredentials+Extension.swift:14]: Are we sure about this? What's to prevent someone setting up UserPools with no IdentityPool, and then calling `toAmplifyAWSCredentials()`?

File: AmplifyPlugins/Core/AWSPluginsCore/Auth/Provider/AuthIdentityProvider.swift [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Amplify
+
+public protocol AuthIdentityProvider {
+    func getIdentityId() -> Result<String, AmplifyAuthError>
+
+    func getUserSub() -> Result<String, AmplifyAuthError>
```
Comment (palpatim) [inline AmplifyPlugins/Core/AWSPluginsCore/Auth/Provider/AuthIdentityProvider.swift:13]: Remind me: "user sub" is a Cognito term? Why isn't our protocol named to indicate this is Cognito-specific?

File: AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthErrorHelper.swift [hunk 10]
```diff
@@ -126,13 +126,12 @@ struct AuthErrorHelper {
...
         // Below error should not happen, these will be handled inside the plugin.
-        case .notSignedIn(let message), // Called in getTokens/getPassword when not signedin to CUP
-        .identityIdUnavailable(let message), // From getIdentityId. Handled in plugin
-        .guestAccessNotAllowed(let message), // Returned from getAWSCredentials. Handled in plugin
+        case .notSignedIn(let message), // HANDLED Called in getTokens/getPassword when not signedin to CUP
+        .identityIdUnavailable(let message), // HANDLED From getIdentityId. Handled in plugin
+        .guestAccessNotAllowed(let message), // HANDLED Returned from getAWSCredentials. Handled in plugin
         .federationProviderExists(let message), // User is already signed in to user pool in federatedSignIn
-        .unableToSignIn(let message), // Called in signout, releaseSignInWait.
+        .unableToSignIn(let message), // HANDLED Called in signout, releaseSignInWait.
```
Comment (palpatim) [inline AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthErrorHelper.swift:133]: I assume you'll be removing these "HANDLED" tags?

File: AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthCognitoSignedOutSessionHelper.swift [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+import Amplify
+import AWSMobileClient
+import AWSPluginsCore
+
+struct AuthCognitoSignedOutSessionHelper {
+
+    /// Creates a signedOut session information with valid identityId and aws credentials.
+    /// - Parameters:
+    ///   - identityId: Valid identity id for the current signedOut session
+    ///   - awsCredentials: Valid AWS Credentials for the current signedOut session
```
Comment (palpatim) [inline AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthCognitoSignedOutSessionHelper.swift:17]: I thought we had agreed that the "signed out" state wouldn't return credentials?

File: AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthCognitoSignedOutSessionHelper.swift [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+        return authSession
+    }
+
+    /// Creates a signedOut session by handling the error
+    /// - Parameter error: Error received while fetching session information
+    /// - Returns: SignedOut session with isSignedIn = false and appropriate error result for other values.
+    static func signedOutSession(withError error: Error) -> AWSAuthCognitoSession {
+
+        if let urlError = error as NSError?, urlError.domain == NSURLErrorDomain {
+            return offlineSignedOutSession()
```
Comment (palpatim) [inline AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthCognitoSignedOutSessionHelper.swift:37]: nit: here and elsewhere: name these factory methods with `make` to clarify that they're creating something, not doing some side effect. If I'm mistaken and the methods are also changing system state, then reflect that in the name too, or break the side effect into a different method

File: AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthorizationProviderAdapter+SignedInSession.swift (hunk 2 above)
Comment (royjit) [reply AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthorizationProviderAdapter+SignedInSession.swift]: Moved few logics to different methods.

File: AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthorizationProviderAdapter+SignedInSession.swift (hunk 3 above)
Comment (royjit) [reply AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthorizationProviderAdapter+SignedInSession.swift:35]: Added

File: AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthorizationProviderAdapter+SignedInSession.swift (hunk 5 above)
Comment (royjit) [reply AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthorizationProviderAdapter+SignedInSession.swift]: Added, let me know if the message is ok..
